src/tools/rbd_mirror/image_replayer/PrepareRemoteImageRequest.cc

```diff
@@ -114,11 +114,7 @@ void PrepareRemoteImageRequest<I>::handle_get_mirror_info(int r) {
     dout(5) << "remote image mirroring is being disabled" << dendl;
     finish(-ENOENT);
     return;
-  } else if (m_promotion_state != librbd::mirror::PROMOTION_STATE_PRIMARY &&
-             (state_builder == nullptr ||
-              state_builder->local_image_id.empty() ||
-              state_builder->local_promotion_state ==
-                librbd::mirror::PROMOTION_STATE_UNKNOWN)) {
+  } else if (m_promotion_state != librbd::mirror::PROMOTION_STATE_PRIMARY) {
     // no local image and remote isn't primary -- don't sync it
     dout(5) << "remote image is not primary -- not syncing" << dendl;
```

Comment:
@MrFreezeex I think I have found why it does not work after failovering now.

I think previously, for not primary remote image the `PrepareRemoteImageRequest` still proceeded and registered the client with the starting (initial) journal position. But after your change the client registration is skipped. The client is registered only after the image is promoted, with the not the starting position but the current position, so when later it lists the tags, the older tags are skipped as "already committed" and it can't detect the primary state properly and reports split-brain.

Not sure I was clear here. If you need more details just let me know.

Reply:
Oh ok thank you for discovering this, I wasn't suspecting at all that this change could produce the bug... IIRC I did this because I thought this was an extra check for safetyness that probably prevented something for multipeer (at least at the time I did this). Will check what is it exactly and find a valid workaround!

Reply:
Actually I think the current rbd-mirror behaviour is buggy and I created a ticket for this bug [1]. Eventually we will need to improve this. And yes, right now you may just try to find a workaround so it would work for your multipeer implementation similarly how it works for the current (single peer) implementation (i.e. not breaking the current promote/demote tests).

Or if you wish you could try to provide a fix for [1] as a separate PR (I am not sure when I have time to work on it) and hopefully you will not need a workaround in this PR then :-)

[1] https://tracker.ceph.com/issues/53460

Reply:
I looked at the code and apparently there is an exception for the journal in journal/PrepareReplayRequest https://github.com/ceph/ceph/blob/master/src/tools/rbd_mirror/image_replayer/journal/PrepareReplayRequest.cc#L71-L80.

If I revert this part of the patch to the original code and I test to create an image and demote the primary image, this condition in PrepareRemoteImageRequest is not triggered and the BootstrapRequest fails later on (even for snapshots). I don't know what I changed in the code to have this behavior, But since the condition doesn't seems right I don't know if I should really debug that more deeply...

Maybe the new condition should be something like `m_promotion_state != librbd::mirror::PROMOTION_STATE_PRIMARY && (m_promotion_state != librbd::mirror::PROMOTION_STATE_ORPHAN || last_tag_synced)`. So that would an exception to replay the demote tag except if we already synced the last tag assuming we can get the tags of both images at this stage and compare the last tags; didn't checked this yet... Before I go ahead with the implementation, does this seems right to you?

Reply:
I am not quite sure I got you. 

Anyway, the first question is why did you need to change this part of the code in the first place (i.e. removing the check for this additional condition `local_promotion_state == PROMOTION_STATE_UNKNOWN`)? This condition seems reasonable to me and makes the bootstrap process not to quit immediately if the local state unknown and proceed with registering journal client, which makes it detect demotion/promotion events later.

When you say "If I revert this part" do you mean this condition in `PrepareRemoteImageRequest<I>::handle_get_mirror_info` or the condition in `PrepareReplayRequest<I>::send` you referred? If this is the first then I expect the bootstrap would quit (fail) later, but is this a problem?

And I don't quite understand your idea with adding `(m_promotion_state != librbd::mirror::PROMOTION_STATE_ORPHAN || last_tag_synced)` condition, i.e. how it should help with replaying already synced tags. If we the journal position is at some tag, the cls list tags methods does not return older tags, i.e. the rbd-mirror does not see them. Or what do you mean?

[1] https://tracker.ceph.com/issues/53460

Reply:
> Anyway, the first question is why did you need to change this part of the code in the first place (i.e. removing the check for this additional condition local_promotion_state == PROMOTION_STATE_UNKNOWN)? This condition seems reasonable to me and makes the bootstrap process not to quit immediately if the local state unknown and proceed with registering journal client, which makes it detect demotion/promotion events later.

I'm not sure, checking...

> If this is the first then I expect the bootstrap would quit (fail) later, but is this a problem?
It it the first, for the second peer it would need to fix the error that is returned so that it's not a generic bootstrap error. In mutipeer this is problematic because PrepareRemoteImageRequest is supposed to detect the primary image between all the peers... A bad detection could prevent the replication.


> And I don't quite understand your idea with adding (m_promotion_state != librbd::mirror::PROMOTION_STATE_ORPHAN || last_tag_synced) condition, i.e. how it should help with replaying already synced tags. If we the journal position is at some tag, the cls list tags methods does not return older tags, i.e. the rbd-mirror does not see them. Or what do you mean?

> [1] https://tracker.ceph.com/issues/53460
 
This is to replace the condition in `PrepareRemoteImageRequest.cc` so this is a failure condition => If this image is in state orphan and that we have not yet replayed the orphan tag we should replay (at least I think?). The check on the tags would probably only help in multipeer to not select the same orphaned image multiple times. Does it make more sense? Maybe I should add the check on PROMOTION_STATE_UKNOWN also, not sure...

Reply:
Ah well I now remember, changing this condition is actually directly linked to the change in BootstrapRequest for multipeer. I need to detect that the image is not the primary early on to select the correct peer. For instance, if local image is non primary and we have a non primary peer first and then a primary one, if I don't change this condition it would match the first non primary peer (as having both peer as non primary don't trigger the condition and thus make the PrepareRemoteImageRequest succeeds without any error).

Reply:
The first solution that comes in my mind is that you could modify your change that in the case of `m_promotion_state != librbd::mirror::PROMOTION_STATE_PRIMARY`, instead of quitting immediately you save the `-EREMOTEIO` return code and proceed with registering the journal client if needed, and then exit with the saved code.

Without trying I have no idea if this is good enough (will not introduce some undesirable effects), and in long term it might be just a workaround, because we will still need to fix the issue with demotion/promotion when the rbd-mirror is not running. But I think you could try this if you don't immediately see an issue with it.

Reply:
Seems to work as a charm with 2 clusters. With 3 clusters, things get a little more complicated apparently. I got:
`2021-12-06T11:17:23.455+0000 7fe86103c700 -1 rbd::mirror::image_replayer::journal::PrepareReplayRequest: 0x5643e5a46f00 handle_get_remote_tags: split-brain detected -- skipping image replay`. I have all the tags printed in here so I believe I should be able to fix that (hopefully).

I will try to take a look at the rbd-mirror promotion/demotion when the daemon is offline and write a test for it, but as you may have noticed I am not familiar at all with the journal logic so no guarantee that I am able to fix it.

Reply:
So the "offline" problem is that if the rbd-mirror is not started it does not register itself in the journal, and the journal may be trimmed (it is trimmed to the min registered client commit position, and if no clients are registered then there is no restrictions). And when after demote/promote the rbd-mirror is started and it registers the journal client it uses the current position and tag list does not show previous demotion tags.

Right now I am not sure about a solution, but I would look at these:
- At the moment when client is registered (`journal_client_register` from cls/cls_journal/cls_journal.cc) it calls `find_min_commit_position` to find the position to register at. I would check if it indeeds finds the min position, i.e. a lower position with demotion entry is already trimmed from the journal. If not then it is probably just need to be fixed.
- If the journal indeed trimmed so we can't register at a lower position then may be we have to change journal/JournalTrimmer.cc not to be so aggressive when trimming if the entry to trim is demote (we don't expect more writes after this any more).

I have not looked at the code closely so I may be completely wrong with these suggestions. Still I hope it provides some tips for you where to look. I hope if will have time to look at it later this week if you still need some help.